libdd-profiling/src/internal/profile/profiles_dictionary_translator.rs

```diff
@@ -0,0 +1,180 @@
+                // this method.
+                let function = unsafe { *self.profiles_dictionary.functions().get(set_id) };
+                // SAFETY: safe if the strings were made by
+                // `self.profiles_dictionary`, which is a precondition of
+                // calling this method.
+                let function = unsafe {
+                    Function {
+                        name: self.translate_string(string_table, function.name)?,
+                        system_name: self.translate_string(string_table, function.system_name)?,
+                        filename: self.translate_string(string_table, function.file_name)?,
+                    }
+                };
+                (Some(set_id), function)
+            }
+        };
+
+        let internal_id = functions
+            .try_dedup(function)
+            .context("failed to deduplicate function in ProfilesDictionaryTranslator")?;
+        self.functions.try_reserve(1).context(
+            "failed to reserve memory for a new function in ProfilesDictionaryTranslator",
+        )?;
+        self.functions.insert(set_id, internal_id);
+        Ok(internal_id)
+    }
+
+    /// Translates a MappingId2 from the ProfilesDictionary into a MappingId
+    /// for this profile's internal collections.
+    ///
+    /// # Safety
+    ///
+    /// The `id2` must have been created by `self.profiles_dictionary`, and
+    /// the strings must also live in the same dictionary.
+    pub unsafe fn translate_mapping(
+        &mut self,
+        mappings: &mut impl Dedup<Mapping>,
+        string_table: &mut StringTable,
+        id2: MappingId2,
+    ) -> anyhow::Result<Option<MappingId>> {
+        // Translate null MappingId2 to Ok(None). This is different from
+        // functions because the internal module uses Option<MappingId>,
+        // whereas it assumes functions are required.
+        let Some(nn) = NonNull::new(id2.0) else {
+            return Ok(None);
+        };
+        let set_id = SetId(nn.cast::<dt::Mapping>());
+        if let Some(internal) = self.mappings.get(&set_id) {
+            return Ok(Some(*internal));
+        }
+
+        // SAFETY: This is safe if `id2` (the MappingId2) was created by
+        // `self.profiles_dictionary`, which is a precondition of calling
+        // this method.
+        let mapping = unsafe { *self.profiles_dictionary.mappings().get(set_id) };
+        let internal = Mapping {
+            memory_start: mapping.memory_start,
+            memory_limit: mapping.memory_limit,
+            file_offset: mapping.file_offset,
+            filename: self.translate_string(string_table, mapping.filename)?,
+            build_id: self.translate_string(string_table, mapping.build_id)?,
+        };
+        let internal_id = mappings
+            .try_dedup(internal)
+            .context("failed to deduplicate mapping in ProfilesDictionaryTranslator")?;
+        self.mappings.try_reserve(1).context(
```

Comment:
Is this going to lead to linear growth in the mappings list instead of logarithmic?

Reply:
No, from [try_reserve's documentation](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.try_reserve):

> Tries to reserve capacity for at least `additional` more elements to be inserted in the HashMap. The collection may reserve more space to speculatively avoid frequent reallocations.